Compilation of µWebSockets has failed and there is no pre-compiled binary available for your system.

### Expected behavior

```
npm install
node app.js
```

starts the app

### Actual behavior

After setting up a node from source, installing via `npm install`, I get the following error when starting the node:

```
$ node app.js 
[dbg] 2018-04-09 23:23:02 | Cache Disabled
(node:3098) UnhandledPromiseRejectionWarning: Error: Compilation of µWebSockets has failed and there is no pre-compiled binary available for your system. Please install a supported C++11 compiler and reinstall the module 'uws'.
    at native (/workspace/lisk/node_modules/uws/uws.js:38:19)
    at Object.<anonymous> (/workspace/lisk/node_modules/uws/uws.js:42:3)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at new SCServer (/workspace/lisk/node_modules/socketcluster-server/scserver.js:101:18)
    at Object.module.exports.attach (/workspace/lisk/node_modules/socketcluster-server/index.js:65:29)
    at /workspace/lisk/node_modules/socketcluster/scworker.js:209:41
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
    at Function.Module.runMain (module.js:695:11)
    at startup (bootstrap_node.js:188:16)
(node:3098) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
```

### Steps to reproduce

### Which version(s) does this affect? (Environment, OS, etc...)

* Ubuntu 18.04 LTS
* Lisk current development branch

---------------

The workaround from https://github.com/SocketCluster/socketcluster/issues/246#issuecomment-373650937 did not help.

In `node_modules/uws/build_log.txt` you can see the error message:

```
make: Entering directory '/workspace/lisk/node_modules/uws/build'
  CXX(target) Release/obj.target/uws/src/Extensions.o
  CXX(target) Release/obj.target/uws/src/Group.o
In file included from ../src/Group.cpp:2:0:
../src/Hub.h:7:10: fatal error: zlib.h: No such file or directory
 #include <zlib.h>
          ^~~~~~~~
compilation terminated.
uws.target.mk:98: recipe for target 'Release/obj.target/uws/src/Group.o' failed
make: *** [Release/obj.target/uws/src/Group.o] Error 1
make: Leaving directory '/workspace/lisk/node_modules/uws/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/share/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Linux 4.15.0-13-generic
gyp ERR! command "/usr/bin/node" "/usr/bin/node-gyp" "rebuild"
gyp ERR! cwd /workspace/lisk/node_modules/uws
gyp ERR! node -v v8.10.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
```

Why does npm install succeed after such an error?

`sudo apt-get install libz-dev && rm -r node_modules/uws && npm install` resolves the error.

Feel free to close this as not-our-bug. Anyway, this is annoying.

--- a/app.js
+++ b/app.js
@@ -473,7 +473,7 @@ d.run(() => {
 						workers: scope.config.wsWorkers,
 						port: scope.config.wsPort,
 						host: '0.0.0.0',
-						wsEngine: scope.config.wsEngine || 'sc-uws',
+						wsEngine: scope.config.wsEngine,
 						appName: 'lisk',
 						workerController: workersControllerPath,
 						perMessageDeflate: false,

--- a/app.js
+++ b/app.js
@@ -473,7 +473,7 @@ d.run(() => {
 						workers: scope.config.wsWorkers,
 						port: scope.config.wsPort,
 						host: '0.0.0.0',
-						wsEngine: scope.config.wsEngine,
+						wsEngine: scope.config.peers.options.wsEngine,
 						appName: 'lisk',
 						workerController: workersControllerPath,
 						perMessageDeflate: false,

--- a/app.js
+++ b/app.js
@@ -473,7 +473,7 @@ d.run(() => {
 						workers: scope.config.wsWorkers,
 						port: scope.config.wsPort,
 						host: '0.0.0.0',
-						wsEngine: 'sc-uws',
+						wsEngine: scope.config.wsEngine || 'sc-uws',
 						appName: 'lisk',
 						workerController: workersControllerPath,
 						perMessageDeflate: false,
@@ -522,6 +522,11 @@ d.run(() => {
 					// The 'fail' event aggregates errors from all SocketCluster processes.
 					scope.socketCluster.on('fail', err => {
 						scope.logger.error(err);
+						if (err.name === 'WSEngineInitError') {
+							var extendedError =
+								'If you are not able to install sc-uws, you can try setting the wsEngine property in config.json to "ws" before starting the node';
+							scope.logger.error(extendedError);
+						}
 					});
 
 					scope.socketCluster.on('workerExit', workerInfo => {

--- a/config/alphanet/config.json
+++ b/config/alphanet/config.json
@@ -62,7 +62,8 @@
 		},
 		"options": {
 			"timeout": 5000,
-			"broadhashConsensusCalculationInterval": 5000
+			"broadhashConsensusCalculationInterval": 5000,
+			"wsEngine": "sc-uws"
 		}
 	},
 	"broadcasts": {

--- a/config/betanet/config.json
+++ b/config/betanet/config.json
@@ -78,7 +78,8 @@
 		},
 		"options": {
 			"timeout": 5000,
-			"broadhashConsensusCalculationInterval": 5000
+			"broadhashConsensusCalculationInterval": 5000,
+			"wsEngine": "sc-uws"
 		}
 	},
 	"broadcasts": {

--- a/config/devnet/config.json
+++ b/config/devnet/config.json
@@ -62,7 +62,8 @@
 		},
 		"options": {
 			"timeout": 5000,
-			"broadhashConsensusCalculationInterval": 5000
+			"broadhashConsensusCalculationInterval": 5000,
+			"wsEngine": "sc-uws"
 		}
 	},
 	"broadcasts": {

--- a/config/mainnet/config.json
+++ b/config/mainnet/config.json
@@ -98,7 +98,8 @@
 		},
 		"options": {
 			"timeout": 5000,
-			"broadhashConsensusCalculationInterval": 5000
+			"broadhashConsensusCalculationInterval": 5000,
+			"wsEngine": "sc-uws"
 		}
 	},
 	"broadcasts": {

--- a/config/testnet/config.json
+++ b/config/testnet/config.json
@@ -78,7 +78,8 @@
 		},
 		"options": {
 			"timeout": 5000,
-			"broadhashConsensusCalculationInterval": 5000
+			"broadhashConsensusCalculationInterval": 5000,
+			"wsEngine": "sc-uws"
 		}
 	},
 	"broadcasts": {

--- a/helpers/config.js
+++ b/helpers/config.js
@@ -146,6 +146,10 @@ function Config(packageJson) {
 		appConfig.coverage = true;
 	}
 
+	if (appConfig.wsEngine === undefined || appConfig.wsEngine === null) {
+		appConfig.wsEngine = 'sc-uws';
+	}
+
 	if (
 		appConfig.api.options.cors.origin === undefined ||
 		appConfig.api.options.cors.origin === null

--- a/helpers/config.js
+++ b/helpers/config.js
@@ -146,8 +146,11 @@ function Config(packageJson) {
 		appConfig.coverage = true;
 	}
 
-	if (appConfig.wsEngine === undefined || appConfig.wsEngine === null) {
-		appConfig.wsEngine = 'sc-uws';
+	if (
+		appConfig.peers.options.wsEngine === undefined ||
+		appConfig.peers.options.wsEngine === null
+	) {
+		appConfig.peers.options.wsEngine = 'sc-uws';
 	}
 
 	if (

--- a/schema/config.js
+++ b/schema/config.js
@@ -213,6 +213,12 @@ module.exports = {
 							timeout: {
 								type: 'integer',
 							},
+							broadhashConsensusCalculationInterval: {
+								type: 'integer',
+							},
+							wsEngine: {
+								type: 'string',
+							},
 						},
 						required: ['timeout'],
 					},

